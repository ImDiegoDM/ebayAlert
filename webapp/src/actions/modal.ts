export function modalOpen(onConfirm: () => void, text: string) {
  return {
    payload: {
      onConfirm,
      text,
    },
    type: 'MODAL_OPEN',
  };
}

export function modalClose() {
  return {
    type: 'MODAL_CLOSE',
  };
}

export function modalConfirm() {
  return {
    type: 'MODAL_CONFIRM',
  };
}
