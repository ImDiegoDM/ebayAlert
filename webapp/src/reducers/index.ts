import { combineReducers } from 'redux';
import searchPhrases from './search_phrases';

export default combineReducers({
  searchPhrases,
});
