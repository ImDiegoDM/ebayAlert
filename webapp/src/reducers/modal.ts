import { AxiosError, AxiosResponse } from 'axios';
import { Action } from 'redux';
import { IApiRequest, ISearchPhrase } from '../interfaces';

export interface IModal {
  open: boolean;
  text?: string;
  onConfirm?: () => void;
}

const defaultValue: IModal = {
  open: false,
};

interface IAction extends Action {
  payload: {
    onConfirm?: () => void;
    text: string;
  };
}

export default (state = defaultValue, action: IAction) => {
  switch (action.type) {
    case 'MODAL_OPEN':
      return {...state, open: true, onConfirm: action.payload.onConfirm, text: action.payload.text};
    case 'MODAL_CLOSE':
      return { ...state, open: false, onConfirm: undefined };
    case 'MODAL_CONFIRM':
      if (state.onConfirm) {
        state.onConfirm();
      }

      return { ...state, open: false, onConfirm: undefined };
    default:
      return state;
  }
};
