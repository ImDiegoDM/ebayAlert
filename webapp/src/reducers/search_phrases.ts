import { AxiosError, AxiosResponse } from 'axios';
import { Action } from 'redux';
import { IApiRequest, ISearchPhrase } from '../interfaces';

const defaultValue: IApiRequest<ISearchPhrase[]> = {
  data: undefined,
  fetched: false,
  fetching: false,
  ok: false,
};

interface IAction extends Action {
  payload: AxiosResponse<ISearchPhrase[]>|AxiosError;
}

export default (state = defaultValue, action: IAction) => {
  switch (action.type) {
    case 'FETCH_PHRASES_PENDING':
      return {...state, fetching: true};
    case 'FETCH_PHRASES_FULFILLED':
      return {
        ...state,
        data: (action.payload as AxiosResponse<ISearchPhrase[]>).data,
        fetched: true ,
        fetching: false,
        ok: true,
      };
    case 'FETCH_PHRASES_REJECTED':
      return {
        ...state,
        error: action.payload as AxiosError,
        fetched: true,
        fetching: false ,
        ok: false,
      };
    default:
      return state;
  }
};
