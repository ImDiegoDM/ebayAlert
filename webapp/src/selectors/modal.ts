import { State } from '../reducers';

export function selectModal(store: State) {
  return store.modal;
}
