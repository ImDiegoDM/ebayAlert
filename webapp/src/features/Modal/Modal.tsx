import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components';
import { modalClose, modalConfirm } from '../../actions/modal';
import { selectModal } from '../../selectors/modal';

const Container = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color:rgba(0, 0, 0, 0.51);
  z-index: 20;
`;

const ModalBox = styled.div`
  width: 450px;
  height: 300px;
  display:flex;
  flex-direction: column;
  align-items: center;
  background-color:white;
  border-radius: 10px;
  position:absolute;
  top: 50px;
  left: 50%;
  transform: translateX(-50%);
  justify-content: space-evenly;
  text-align: center;
  padding: 20px;
`;

const Buttons = styled.div`
  display: flex;
`;

const Button = styled.button`
  margin-right: 15px;
`;

export function Modal() {
  const modalState = useSelector(selectModal);
  const dispatch = useDispatch();

  if (!modalState.open) {
    return <></>;
  }

  return <Container>
    <ModalBox>
      <h2>{modalState.text}</h2>
      <Buttons>
        <Button onClick={() => dispatch(modalClose())}>Cancel</Button>
        <Button onClick={() => dispatch(modalConfirm())}>Confirm</Button>
      </Buttons>
    </ModalBox>
  </Container>;
}
