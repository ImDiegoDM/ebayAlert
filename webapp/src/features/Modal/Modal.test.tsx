import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import promise from 'redux-promise-middleware';
import reducer from '../../reducers';
import { IModal } from '../../reducers/modal';
import { Modal } from './Modal';

const modalOpen: IModal = {
  open: true,
  text: 'Modal Text',
};

const modalClosed: IModal = {
  open: false,
};

describe('<Error/>', () => {
  it('should match snapshot when the store state is open', () => {

    const middlware = applyMiddleware(promise);

    const store = createStore(reducer, {modal: modalOpen}, middlware);

    const { container } = render(<Provider store={store}>
      <Modal/>
    </Provider>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when the store state is closed', () => {

    const middlware = applyMiddleware(promise);

    const store = createStore(reducer, {modal: modalClosed}, middlware);

    const { container } = render(<Provider store={store}>
      <Modal/>
    </Provider>);

    expect(container.firstChild).toMatchSnapshot();
  });
});
