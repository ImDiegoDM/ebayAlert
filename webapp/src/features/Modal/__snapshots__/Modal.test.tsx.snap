// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Error/> should match snapshot when the store state is closed 1`] = `null`;

exports[`<Error/> should match snapshot when the store state is open 1`] = `
.c0 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0,0,0,0.51);
  z-index: 20;
}

.c1 {
  width: 450px;
  height: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: white;
  border-radius: 10px;
  position: absolute;
  top: 50px;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  text-align: center;
  padding: 20px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  margin-right: 15px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <h2>
      Modal Text
    </h2>
    <div
      class="c2"
    >
      <button
        class="c3"
      >
        Cancel
      </button>
      <button
        class="c3"
      >
        Confirm
      </button>
    </div>
  </div>
</div>
`;
